File: go/bottles.go
package bottles

import "fmt"

func Verse(beers int) string {
  result := ""

  switch {
  case beers > 2:
    result = fmt.Sprintf("%d bottles of beer on the wall, %d bottles of beer.\nTake one down and pass it around, %d bottles of beer on the wall.\n", beers, beers, beers - 1)
  case beers > 1:
    result = fmt.Sprintf("%d bottles of beer on the wall, %d bottles of beer.\nTake one down and pass it around, %d bottle of beer on the wall.\n", beers, beers, beers - 1)
  case beers > 0:
    result = fmt.Sprintf("%d bottle of beer on the wall, %d bottle of beer.\nTake it down and pass it around, no more bottles of beer on the wall.\n", beers, beers)
  default:
    result = fmt.Sprintf("No more bottles of beer on the wall, no more bottles of beer.\nGo to the store and buy some more, 99 bottles of beer on the wall.\n")
  }

  return result
}

func Verses(starting int, ending int) string {
  result := ""

  for beers := starting; beers >= ending; beers-- {
    result += Verse(beers)

    if beers != ending {
      result += "\n"
    }
  }

  return result
}

func Song() string {
  result := Verses(99, 0)

  return result
}

File: ruby/lib/bottles.rb
class Bottles
  def song
    verses(99, 0)
  end

  def verses(start, finish)
    start.downto(finish).map { |beers| verse(beers) }.join("\n")
  end

  def verse(beers)
    case beers
    when 0
      "No more bottles of beer on the wall, no more bottles of beer.\n"\
      "Go to the store and buy some more, 99 bottles of beer on the wall.\n"
    when 1
      "#{beers} bottle of beer on the wall, #{beers} bottle of beer.\n"\
      "Take it down and pass it around, no more bottles of beer on the wall.\n"
    when 2
      "#{beers} bottles of beer on the wall, #{beers} bottles of beer.\n"\
      "Take one down and pass it around, #{beers-1} bottle of beer on the wall.\n"
    else
      "#{beers} bottles of beer on the wall, #{beers} bottles of beer.\n"\
      "Take one down and pass it around, #{beers-1} bottles of beer on the wall.\n"
    end
  end
end
